presto-main/src/main/java/com/facebook/presto/operator/PagesIndex.java

```diff
@@ -437,6 +441,30 @@ public LookupSourceSupplier createLookupSourceSupplier(
         return createLookupSourceSupplier(session, joinChannels, hashChannel, filterFunctionFactory, sortChannel, searchFunctionFactories, Optional.empty());
     }
 
+    public PagesSpatialIndexSupplier createPagesSpatialIndex(
+            Session session,
+            int geometryChannel,
+            BiPredicate<OGCGeometry, OGCGeometry> spatialRelationshipTest,
+            Optional<JoinFilterFunctionFactory> filterFunctionFactory,
+            List<Integer> outputChannels)
+    {
+        // TODO probably shouldn't copy to reduce memory and for memory accounting's sake
+        List<List<Block>> channels = ImmutableList.copyOf(this.channels);
+        return new PagesSpatialIndexSupplier(session, valueAddresses, types, outputChannels, channels, sizeOfChannels(), geometryChannel, spatialRelationshipTest, filterFunctionFactory);
+    }
+
+    private long sizeOfChannels()
+    {
+        return IMMUTABLE_LIST_HEADER_SIZE +
+                stream(channels)
```

Comment:
@highker this is different than @nezihyigitbasi 's https://github.com/prestodb/presto/commit/bd06b0b05fdef05349d605ae18bb3e5529f595a7, because AFAIU this is called once in operator's lifetime